File: coding-labs/lru-ttl-cache/java/src/main/java/LruTtlCache.java
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;
import java.util.function.LongSupplier;

public final class LruTtlCache<K, V> {
  private record Entry<V>(V value, long expiresAt) {}
  private final int capacity; private final LongSupplier clock;
  private final LinkedHashMap<K, Entry<V>> entries = new LinkedHashMap<>(16, 0.75f, true);
  public LruTtlCache(int capacity, LongSupplier clock) { if (capacity <= 0) throw new IllegalArgumentException("capacity must be positive"); this.capacity = capacity; this.clock = clock; }
  public synchronized void put(K key, V value, long ttlMs) {
    if (ttlMs < 0) throw new IllegalArgumentException("ttl must be non-negative");
    entries.put(key, new Entry<>(value, clock.getAsLong() + ttlMs));
    while (entries.size() > capacity) entries.remove(entries.keySet().iterator().next());
  }
  public synchronized Optional<V> get(K key) {
    Entry<V> entry = entries.get(key);
    if (entry == null) return Optional.empty();
    if (clock.getAsLong() >= entry.expiresAt()) { entries.remove(key); return Optional.empty(); }
    return Optional.ofNullable(entry.value());
  }
  public synchronized int size() { return entries.size(); }
}

File: coding-labs/lru-ttl-cache/python/src/lru_ttl_cache.py
from collections import OrderedDict

class LruTtlCache:
    def __init__(self, capacity, now_ms):
        if capacity <= 0: raise ValueError("capacity must be positive")
        self._capacity, self._now, self._entries = capacity, now_ms, OrderedDict()

    def put(self, key, value, ttl_ms):
        if ttl_ms < 0: raise ValueError("ttl must be non-negative")
        if key is None or value is None: raise ValueError("key and value are required")
        self._entries.pop(key, None)
        self._entries[key] = (value, self._now() + ttl_ms)
        while len(self._entries) > self._capacity: self._entries.popitem(last=False)

    def get(self, key):
        if key is None: raise ValueError("key is required")
        entry = self._entries.get(key)
        if entry is None: return None
        if self._now() >= entry[1]: self._entries.pop(key); return None
        self._entries.move_to_end(key)
        return entry[0]

    def size(self): return len(self._entries)
